mmdet/models/dense_heads/yolo_head.py

```diff
@@ -0,0 +1,495 @@
+        num_imgs = len(img_metas)
+        device = pred_maps[0][0].device
+
+        featmap_sizes = [
+            pred_maps[i].shape[-2:] for i in range(self.num_levels)
+        ]
+        multi_level_anchors = self.anchor_generator.grid_anchors(
+            featmap_sizes, device)
+        anchor_list = [multi_level_anchors for _ in range(num_imgs)]
+
+        responsible_flag_list = []
+        for img_id in range(len(img_metas)):
+            responsible_flag_list.append(
+                self.anchor_generator.responsible_flags(
+                    featmap_sizes, gt_bboxes[img_id], device))
+
+        target_maps_list, neg_maps_list = self.get_targets(
+            anchor_list, responsible_flag_list, gt_bboxes, gt_labels)
+
+        losses_cls, losses_conf, losses_xy, losses_wh = multi_apply(
+            self.loss_single, pred_maps, target_maps_list, neg_maps_list)
+
+        return dict(
+            loss_cls=losses_cls,
+            loss_conf=losses_conf,
+            loss_xy=losses_xy,
+            loss_wh=losses_wh)
+
+    def loss_single(self, pred_map, target_map, neg_map):
+        """Compute loss of a single image from a batch.
+
+        Args:
+            pred_map (Tensor): Raw predictions for a single level.
+            target_map (Tensor): The Ground-Truth target for a single level.
+            neg_map (Tensor): The negative masks for a single level.
+
+        Returns:
+            tuple:
+                loss_cls (Tensor): Classification loss.
+                loss_conf (Tensor): Confidence loss.
+                loss_xy (Tensor): Regression loss of x, y coordinate.
+                loss_wh (Tensor): Regression loss of w, h coordinate.
+        """
+
+        num_imgs = len(pred_map)
+        pred_map = pred_map.permute(0, 2, 3,
+                                    1).reshape(num_imgs, -1, self.num_attrib)
+        neg_mask = neg_map.float()
+        pos_mask = target_map[..., 4]
+        pos_and_neg_mask = neg_mask + pos_mask
+        pos_mask = pos_mask.unsqueeze(dim=-1)
+        if torch.max(pos_and_neg_mask) > 1.:
+            warnings.warn('There is overlap between pos and neg sample.')
+            pos_and_neg_mask = pos_and_neg_mask.clamp(min=0., max=1.)
+
+        pred_xy = pred_map[..., :2]
+        pred_wh = pred_map[..., 2:4]
+        pred_conf = pred_map[..., 4]
+        pred_label = pred_map[..., 5:]
+
+        target_xy = target_map[..., :2]
+        target_wh = target_map[..., 2:4]
+        target_conf = target_map[..., 4]
+        target_label = target_map[..., 5:]
+
+        if self.balance_conf:
+            num_pos_gt = max(int(torch.sum(target_conf)), 1)
+            num_total_grids = target_conf.numel()
+            pos_weight = num_total_grids / num_pos_gt
+            conf_loss_weight = 1 / pos_weight
+        else:
+            pos_weight = 1
+            conf_loss_weight = 1
+
+        pos_weight = target_label.new_tensor(pos_weight)
+
+        loss_cls = F.binary_cross_entropy_with_logits(
```

Comment:
Could we use `build_loss` to use loss modules written in mmdet.models.losses as those in other heads?

Reply:
Hi @xvjiarui , 

First, thanks for your work in the past several weeks! Please let me know if I can help.

Speaking of the loss functions, the reduction method is worth noticing. In gluno-cv and ultralytics' implementation, the reduction method is `mean` (see https://github.com/ultralytics/yolov3/blob/bf34ae007fea165255c8be1b7a835dedc49a9613/utils/utils.py#L358), while in this repo, it's `sum`. It's surprising to see that this does not make much difference on `mAP`. What's your opinion? Thanks.

Reply:
Hi @ElectronicElephant 
The sum is based on the original implementation of this PR. 
I just refactored loss function:
1. Use `build_loss`
2. Replace `sum` with `mean`. 
3. Remove `grad_clip`. 
The new experiments are expected to finish in 3 days.

Reply:
Hi @xvjiarui,
I think the loss you built is not properly averaged. When take averages, the averaging factor should be the number of positive samples for xy, wh, cls and the total number of positive and negative samples for conf (possibly divided by the number of classses as well). And just a reminder, the config file you commit is different from the previous version for the weights of xy and wh (by a factor of 2).
Regards.

Reply:
Hi @sudo-rm-covid19
I reverted my modifications. I find that using `avg` yields lower performance (maybe some other hyperparameter needs tuning). And remove `grad_clip` will make the training quite unstable. 
So I just refactored loss function and keep the behavior the same as before.